src/borrow.rs

```diff
@@ -167,88 +241,150 @@ impl BorrowFlags {
     }
 
     #[allow(clippy::mut_from_ref)]
-    unsafe fn get(&self) -> &mut HashMap<usize, isize> {
+    unsafe fn get(&self) -> &mut BorrowFlagsInner {
         (*self.0.get()).get_or_insert_with(HashMap::new)
     }
 
-    fn acquire<T, D>(&self, array: &PyArray<T, D>) -> Result<(), BorrowError> {
+    fn acquire<T, D>(&self, array: &PyArray<T, D>) -> Result<(), BorrowError>
+    where
+        T: Element,
+        D: Dimension,
+    {
         let address = base_address(array);
+        let key = BorrowKey::from_array(array);
 
         // SAFETY: Access to `&PyArray<T, D>` implies holding the GIL
         // and we are not calling into user code which might re-enter this function.
         let borrow_flags = unsafe { BORROW_FLAGS.get() };
 
         match borrow_flags.entry(address) {
             Entry::Occupied(entry) => {
-                let readers = entry.into_mut();
-
-                let new_readers = readers.wrapping_add(1);
-
-                if new_readers <= 0 {
-                    cold();
-                    return Err(BorrowError::AlreadyBorrowed);
+                let same_base = entry.into_mut();
+
+                if let Some(readers) = same_base.get_mut(&key) {
+                    // Zero flags are removed during release.
+                    assert_ne!(*readers, 0);
+
+                    let new_readers = readers.wrapping_add(1);
+
+                    if new_readers <= 0 {
+                        cold();
```

Comment:
Not sure, but in this case I certainly want to pessimize the failure case because calling this method and reaching failure will never be part of a working program. Hence, if there is any codegen leeway to be hand by doing so, I think the compiler should get it.